File: Mile Pace.java
// Time Complexity: O(1) - all operations are constant time
public class MilePace {
    public static String milePace(double miles, String timeStr) {
        // Step 1: convert time string "MM:SS" to total seconds
        String[] parts = timeStr.split(":");
        int minutes = Integer.parseInt(parts[0]);
        int seconds = Integer.parseInt(parts[1]);
        int totalSeconds = minutes * 60 + seconds;

        // Step 2: average seconds per mile
        double paceSeconds = totalSeconds / miles;

        // Step 3: convert back to MM:SS
        int paceMinutes = (int) (paceSeconds / 60);
        int paceRemainingSeconds = (int) Math.round(paceSeconds % 60);

        // Handle rounding edge case (e.g. 59.9 sec -> 60 sec)
        if (paceRemainingSeconds == 60) {
            paceMinutes += 1;
            paceRemainingSeconds = 0;
        }

        return String.format("%02d:%02d", paceMinutes, paceRemainingSeconds);
    }
}

File: Mile Pace.py
# Time Complexity: O(1) - all operations are constant time
# Space Complexity: O(1) - only a fixed number of variables used
def mile_pace(miles: float, time_str: str) -> str:
	# Split the time string into minutes and seconds
	minutes, seconds = map(int, time_str.split(":"))
	# Convert minutes and seconds to total seconds
	total_seconds = minutes * 60 + seconds

	# Calculate average seconds per mile
	pace_seconds = total_seconds / miles

	# Convert pace in seconds back to minutes and seconds
	pace_minutes = int(pace_seconds // 60)
	pace_remaining_seconds = int(round(pace_seconds % 60))

	# If rounding results in 60 seconds, increment minutes and reset seconds
	if pace_remaining_seconds == 60:
		pace_minutes += 1
		pace_remaining_seconds = 0

	# Format the result as MM:SS string
	return f"{pace_minutes:02d}:{pace_remaining_seconds:02d}"

# This function calculates the average pace per mile given the total distance (in miles)
# and the total time (in MM:SS format). It returns the pace as a string in MM:SS format.
